File: src/C3.Infrastructure/Updates/UpdateReleaseManifest.vb
Namespace Updates

    Public NotInheritable Class UpdateChecksumManifest

        Friend Sub New(
                fileName As String,
                length As Long,
                sha256 As String,
                url As String)

            Me.FileName = fileName
            Me.Length = length
            Me.Sha256 = sha256
            Me.Url = url
        End Sub

        Public ReadOnly Property FileName As String

        Public ReadOnly Property Length As Long

        Public ReadOnly Property Sha256 As String

        Public ReadOnly Property Url As String

    End Class

    Public NotInheritable Class UpdateReleasePackage

        Friend Sub New(
                lane As String,
                distribution As String,
                fileName As String,
                length As Long,
                sha256 As String,
                url As String)

            Me.Lane = lane
            Me.Distribution = distribution
            Me.FileName = fileName
            Me.Length = length
            Me.Sha256 = sha256
            Me.Url = url
        End Sub

        Public ReadOnly Property Lane As String

        Public ReadOnly Property Distribution As String

        Public ReadOnly Property FileName As String

        Public ReadOnly Property Length As Long

        Public ReadOnly Property Sha256 As String

        Public ReadOnly Property Url As String

    End Class

    ''' <summary>
    ''' A validated, channel-specific C3 release-feed document.
    ''' </summary>
    Public NotInheritable Class UpdateReleaseManifest

        Friend Sub New(
                channel As String,
                productVersion As String,
                stage As String,
                informationalVersion As String,
                releaseDate As DateTime,
                catalogueWriteFormat As String,
                published As Boolean,
                releaseUrl As String,
                checksumManifest As UpdateChecksumManifest,
                packages As IList(Of UpdateReleasePackage),
                releaseIdentity As SemanticVersion)

            If packages Is Nothing Then
                Throw New ArgumentNullException("packages")
            End If
            Me.Channel = channel
            Me.ProductVersion = productVersion
            Me.Stage = stage
            Me.InformationalVersion = informationalVersion
            Me.ReleaseDate = releaseDate
            Me.CatalogueWriteFormat = catalogueWriteFormat
            Me.Published = published
            Me.ReleaseUrl = releaseUrl
            Me.ChecksumManifest = checksumManifest
            Me.Packages = New List(Of UpdateReleasePackage)(packages).AsReadOnly()
            Me.ReleaseIdentity = releaseIdentity
        End Sub

        Public ReadOnly Property Channel As String

        Public ReadOnly Property ProductVersion As String

        Public ReadOnly Property Stage As String

        Public ReadOnly Property InformationalVersion As String

        Public ReadOnly Property ReleaseDate As DateTime

        Public ReadOnly Property CatalogueWriteFormat As String

        Public ReadOnly Property Published As Boolean

        Public ReadOnly Property ReleaseUrl As String

        Public ReadOnly Property ChecksumManifest As UpdateChecksumManifest

        Public ReadOnly Property Packages As IList(Of UpdateReleasePackage)

        Public ReadOnly Property ReleaseIdentity As SemanticVersion

    End Class

End Namespace

File: src/C3.Infrastructure/Updates/UpdateReleaseManifest.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.Updates
{
	public sealed class UpdateReleaseManifest
	{
		public string Channel
		{
			get;
		}

		public string ProductVersion
		{
			get;
		}

		public string Stage
		{
			get;
		}

		public string InformationalVersion
		{
			get;
		}

		public DateTime ReleaseDate
		{
			get;
		}

		public string CatalogueWriteFormat
		{
			get;
		}

		public bool Published
		{
			get;
		}

		public string ReleaseUrl
		{
			get;
		}

		public UpdateChecksumManifest ChecksumManifest
		{
			get;
		}

		public IList<UpdateReleasePackage> Packages
		{
			get;
		}

		public SemanticVersion ReleaseIdentity
		{
			get;
		}

		internal UpdateReleaseManifest(string channel, string productVersion, string stage, string informationalVersion, DateTime releaseDate, string catalogueWriteFormat, bool published, string releaseUrl, UpdateChecksumManifest checksumManifest, IList<UpdateReleasePackage> packages, SemanticVersion releaseIdentity)
		{
			if (packages == null)
			{
				throw new ArgumentNullException("packages");
			}
			Channel = channel;
			ProductVersion = productVersion;
			Stage = stage;
			InformationalVersion = informationalVersion;
			ReleaseDate = releaseDate;
			CatalogueWriteFormat = catalogueWriteFormat;
			Published = published;
			ReleaseUrl = releaseUrl;
			ChecksumManifest = checksumManifest;
			Packages = new List<UpdateReleasePackage>(packages).AsReadOnly();
			ReleaseIdentity = releaseIdentity;
		}
	}
}
